### PR TITLE
refactor(ivy): make interpolation smaller

### PR DESCRIPTION
**Pull request by benlesh**
**NOTE: This is just a proof of concept**

In this refactor, all `interpolationX` methods are refactored to call the same inner implementation. Realizing that all eight of the previous methods existed to reduce the allocation of arrays that you'd have if you have a single method, this refactor uses a shared array in order to reduce the size of the individual instructions, and have all logic flow through a single function.

### CODE REVIEW

File: packages/core/src/render3/instructions/instructions.ts [hunk 1]
```diff
@@ -2769,6 +2769,82 @@ export function interpolationV(values: any[]): string|NO_CHANGE {
...
+  }
+  return NO_CHANGE;
+}
+
+/**
+ * Sets a single value in `_interpolationParts` and increments the pointer.
+ * @param value the value to set
+ */
+function setInterpolationPart(value: any) {
+  _interpolationParts[_interpolationPartPointer++] = value;
```
Comment (mhevery) [inline packages/core/src/render3/instructions/instructions.ts]: This solution is much slower. :-(

The issue is that it requires multiple reads/writes into memory, where as the existing solution only requires use of the local registers and hence is much faster.

`_interpolationPartPointer++` requires reading memory and than writing into it.
`_interpolationParts[...]` requires writing to memory (and than reading later) It also requires the VM to generate array overflow checks.

**Review by mhevery — APPROVED** [review]
I like the way you think (want to make the code simpler) In this case however, I don't think it will work out.  But please keep thinking about these kinds of ideas. Please run them by me before as it would save you time making a PR.

File: packages/core/src/render3/instructions/instructions.ts (hunk 1 above)
Comment (benlesh) [reply packages/core/src/render3/instructions/instructions.ts]: Makes sense! Thanks!

Comment (benlesh) [reply packages/core/src/render3/instructions/instructions.ts]: Out of curiosity, does it need to check array overflow if the size of the Array is known and less than the result of `_interpolationPointer++` (as in this PR)?

**Comment by benlesh** [comment]
BTW: I wrote the code just to flush out the idea. :man_shrugging:

**Pull request closed without merging**